tainted: mirdeep2: Fix runtime magic patching & add test phase

mschilli87 opened the pull request:
Turns out I missed a dependency when packaging this.

mschilli87 commented:
While I was at it. I also re-based this just in case. :wink:
I think that fix is kinda important as it might break miRDeep2 in a way that is hard to detect for Guix novices at BIMSB. :wink:

rekado commented:
Thank you for the patch!
Sorry for dropping the ball here; this patch came in while I was on parental leave and I forgot about it later...

I think adding "which" to the inputs may not be enough.  We may also need to patch the invocation of "which" to actually use the one from the inputs instead of whatever happens to be in the environment.  The alternative is to add "which" to the propagated-inputs, but I think that's inelegant.

mschilli87 commented:
OK. I think I get it. So I found three uses of `which` during the installation:

* `install.pl`:
  ```perl
  [  67] $grep=`which grep`;
  [  69]   $grep=`which ggrep`;
  [ 104] my $curl=`which curl`;
  ```

And an additional five (potentially) triggered at runtime:

* `src/make_html.pl`:
  ```perl
  [3636]    my $axs=`which miRDeep2.pl`;
  ```
* `src/mapper.pl`:
  ```perl
  [ 746]       my $a=`which miRDeep2.pl`;
  ```
* `src/miRDeep2.pl`:
  ```perl
  [ 149] my $scripts=`which miRDeep2.pl`;
  [1123]   my $a=`which miRDeep2.pl`;
  ```
* `src/quantifier.pl`:
  ```perl
   [1436]    my $a=`which miRDeep2.pl`;
  ```

The ones defining `$a` in `src/mapper.pl`, `src/miRDeep2.pl`, and `src/quantifier.pl` should have no effect though as we already patch the corresponding lines 750, 1126, and 1441, respectively, rendering those variable(s) `$a` (which are only required to define `$bn`) unused:
https://github.com/BIMSBbioinfo/guix-bimsb-nonfree/blob/a8e445259442ff6a2885ca8f51db33e1e4209d29/bimsb/packages/tainted.scm#L188-L196

The other call in `src/miRDeep2.pl` is used to locate `Rfam_for_miRDeep.fa` which should be patched already as well:
https://github.com/BIMSBbioinfo/guix-bimsb-nonfree/blob/a8e445259442ff6a2885ca8f51db33e1e4209d29/bimsb/packages/tainted.scm#L197-L202

That leaves the call in `src/make_html.pl` that is also used to locate installed resources via a hard-coded path relative to the called script, which is at least partially handled already as well:
https://github.com/BIMSBbioinfo/guix-bimsb-nonfree/blob/a8e445259442ff6a2885ca8f51db33e1e4209d29/bimsb/packages/tainted.scm#L203-L208

So is there a Guix way to replace those occurences of `which` by the correct `/gnu/store/.../bin/which`?
Where could I find documention on (or an example of) this?

Does that even make sense at this point? I could double-check if `src/make_html.pl` was changed since I packaged this and needs further patching to not rely on the `which` calls. Then, I assume it would be better practice to actually comment out the lines that depend on `which` to have it not called at all at runtime.

During install, as you can see `which` only used to find dependencies like `grep` and `curl` but I figure we could get rid of these by patching in the appropriate paths in the store instead.

I can try to come up with a better patch but I'd clearly need some more pointers here. :pleading_face:

rekado commented:
You can use `(which "which")` in a build phase to get the store file name for "/bin/which".

mschilli87 commented:
Yeah,I just feel like if we already path the perl scripts we can also patch them such that they don't need `which`. Would `(which "grep")` and `(which "curl")` work for the `install.pl` script? All the runtime 'magic' is only to allow installing miRDeep2 to various places, but in th case of Guix, that place is fixed so there is no real need for `which` at all.

rekado commented:
Yes, `(which "curl")` and `(which "grep")` will work, but you need to make sure the corresponding packages are among the inputs.

I agree with patching out the runtime magic with static guarantees.

mschilli87 commented:
Allright. I'll give it a shot, either later today or tomorrow, and ping you here again.

mschilli87 commented:
OK. I force-pushed a new version that addresses the runtime-calls to `which` by either commenting out the respective lines or replacing the `which` call with the appropriate path in the store.
The `install.pl` script is actually not used by Guix since it pulls dependencies handeled by Guix. So no need to touch it.
Thus, also no need to add `curl` and `grep` as inputs.

Before merging this, I'd like to do two more things that I'd like some feedback for:

There is no `make test` here but miRDeep ships with a subdirectory containing example data and a script to run the pipeline on those. I'd like to replace the test-phase of the `gnu-build-system` by, basically,

```sh
cd src/tutorial && run_tut.sh
```

This way I hope to catch if a future  update requires more/different patching.
How would I go about that? I'd need to be in an environment with the package build and on PATH clearly.
Also, since I did not make `bowtie1` a propagated input, but wrap the installed scripts adjusting PATH, I'd need to do the same for `run_tut.sh` so it can find and run `bowtie-build`.

Finally, while running the entire tutorial manually, I noticed perl complaining about missing `LANGUAGE` and `LC_ALL` which can result in errors durings some checks, probably due to some sorting or regex issues. Anyways, I encountered no issues setting both to `C` and I'd like to make that part of the installation unluess you object.
I assume I could add those variables to the wrapper scripts that already adjust `PATH` and `PERL5LIB`.
But maybe that is a stupid idea hacking around a problem I created by a beginners mistake and you can tell me how the proper solution looks like before I wasteany time on that. ;-)

mschilli87 commented:
Force-pushed again removing some commented out lines I forgot to remove previously. Now it should be ready for review. I did successfully run the tutorial so I am pretty optimistic about the result. But maybe the is a better way to reach it so feel free to comment.

mschilli87 committed:
tainted: mirdeep2: Fix runtime magic patching

miRDeep2 relies on its archive structure to find resources by fixed
paths relative to the executed script.
With Guix installing those to the appropriate places, this always
required some patching. So far, while not relying on running `which`
anymore, the actual call had been left untouched. This breaks miRDeep2
on systems without `which`.
This patch addresses all calls to `which` that may occur when running
miRDeep2 and also addresses further improvements that came up during
reviewing the required changes (with input from Ricardo Wurmus).

mschilli87 commented:
I just re-structured the code a bit for (IMHO) better readability.

rekado commented on bimsb/packages/tainted.scm:
```diff
@@ -196,16 +198,18 @@ deviation (SD) plots, coefficient of variation (CV) plots.")
-                 (("\\$scripts/\\.\\./Rfam_for_miRDeep\\.fa")
-                  (string-append out "/share/mirdeep2/Rfam_for_miRDeep.fa"))))
+                 (("\\$\\{?scripts\\}?/\\.\\.(/Rfam_for_miRDeep\\.fa)" full file)
+                  (string-append out "/share/mirdeep2" file))
```
if you aren't using a match (like `full`) you can write `_` instead.

mschilli87 replied:
I took care of this with the latest push,

rekado commented on bimsb/packages/tainted.scm:
```diff
@@ -196,16 +198,18 @@ deviation (SD) plots, coefficient of variation (CV) plots.")
-               (("mkdir \"\\$\\{scripts\\}indexes")
-                "mkdir -p \"${scripts}indexes"))
+               (("(mkdir) (\"\\$\\{scripts\\}indexes)" call cmd path)
+                (string-append cmd " -p " path)))
```
My ever-growing gut tells me that repeating "mkdir" is preferable to complicating the regular expression with more matching.

There is an alternative to all of this: we could write a proper patch, drop it in the root directory of the repository, and apply it in the `source` field.  We could even come up with a patch that upstream would accept, e.g. running `which` only once and using the result everywhere.  Then the amount of patching we have to maintain would become much smaller.

mschilli87 replied:
Sure, I can replace `cdm "` by  `"mkdir` and drop the extra parents in the pattern. ;-)

I can also write a patch. Just didn't learn how to apply with Guix yet. Do you have an example for how to do that via `source` at hand? Otherwise I'll check the manual myself.  But would prefer that over all the substitutions via scheme. I doubt upstream would expect it though. But feel free to contact @drmirdeep once we have it working here.

mschilli87 replied:
I went with the first solution for now as I don't know when I'll find time to work on a proper patch.
Any improvement is better than none, I guess. ;-)

rekado commented on bimsb/packages/tainted.scm:
```diff
@@ -177,7 +177,7 @@ deviation (SD) plots, coefficient of variation (CV) plots.")
     (build-system gnu-build-system)
     (arguments
-     `(#:tests? #f ;; no check target
+     `(#:tests? #f ;; no check target; TODO: Run tutorial as test?
```
As long as its non-interactive: sure.

mschilli87 replied:
It is.

*edit:* Just to clarify: It is  non-interactive. ;-)

mschilli87 replied:
I'll still need to take care of this before this PR is ready to merge.

mschilli87 replied:
The latest push finally took care of this.

rekado commented on bimsb/packages/tainted.scm:
```diff
@@ -188,6 +188,8 @@ deviation (SD) plots, coefficient of variation (CV) plots.")
                (for-each
```
This line wasn't changed by you, but maybe we can fix this while you're at it.  `substitute*` takes a list of files, so we don't need to `for-each` a `lambda` binding `script` over a list of files --- we can simply write `(substitute* '("src/..." "..." "...")...)`.

mschilli87 replied:
Sure. I'll have a look at it.

mschilli87 replied:
I gave it a shot and it seems I got it right. At least I can still build locally and run the tutorial just fine.
Thanks for the help simplifying this.

rekado commented on bimsb/packages/tainted.scm:
```diff
@@ -1,6 +1,6 @@
 ;;; GNU Guix --- Functional package management for GNU
 ;;; Copyright © 2019, 2020 Ricardo Wurmus <ricardo.wurmus@mdc-berlin.de>
-;;; Copyright © 2019 Marcel Schilling <marcel.schilling@mdc-berlin.de>
+;;; Copyright © 2019, 2020, 2021 Marcel Schilling <marcel.schilling@uni-luebeck.de>
```
Also consider adding a .mailmap file that maps the old email address used in commits to your new address.

mschilli87 replied:
Makes sense. Did I understand correctly that I'd need to explicitely include every old commit? Or is there a 'global'/catchall mapping syntax as well?

rekado replied:
The .mailmap file would contain something like this

```
Marcel Schilling <marcel.schilling@uni-luebeck.de>
Marcel Schilling <marcel.schilling@uni-luebeck.de> <marcel.schilling@mdc-berlin.de>
```

This collates your email addresses, specifying that the uni-luebeck address is the main identity.

mschilli87 replied:
I have added the `.mailcap` file as you suggested with the latest push.

rekado commented:
> I'd like to replace the test-phase of the `gnu-build-system` by, basically,
> 
> ```shell
> cd src/tutorial && run_tut.sh
> ```
> 
> This way I hope to catch if a future update requires more/different patching.
> How would I go about that?

Here's a start:

```scheme
(replace 'check
  (lambda _
    (setenv "PATH" (string-append (getcwd) "/wherever/the/command/is:" (getenv "PATH"))
    (with-directory-excursion "src/tutorial"
      (invoke "run_tut.sh"))))
```


> Finally, while running the entire tutorial manually, I noticed perl complaining about missing LANGUAGE and LC_ALL which can result in errors durings some checks, probably due to some sorting or regex issues. Anyways, I encountered no issues setting both to C and I'd like to make that part of the installation unluess you object.

Yes, that's probably okay.  The correct fix would probably involve adding `glibc-utf8-locales` to the inputs:

```scheme
(lambda* (#:key inputs #:allow-other-keys)
  ...
  (set-path-environment-variable
          "GUIX_LOCPATH" '("lib/locale")
          (list (assoc-ref inputs "glibc-utf8-locales")))
         (setenv "LANG" "en_US.UTF-8")
  ...)
```

Users should have installed `glibc-locales` or `glibc-utf8-locales` and set GUIX_LOCPATH anyway, so you shouldn't hardcode that.

mschilli87 commented:
Thx for the helpful pointers. I am not sure I understand what 'the command' in `/wherever/the/command/is:` is and where the lambda using `GUIX_LOCPATH` goes (or how it even works) but I can spend some time on that myself and udate this PR for another round of review by you. I doubt I'll get to it this week though.

rekado commented:
Sorry for not being clearer.

> I am not sure I understand what 'the command' in `/wherever/the/command/is`:

You want to run a program in the `check` phase.  I don't know what that program does exactly, and what executables it expects to find on PATH.  My suggestion was just to add the directories containing those executables to PATH.

> where the lambda using GUIX_LOCPATH goes

If you just want to fix locales during the execution of the tutorial you only need it in the `check` phase.

So your whole check phase would be a combination of the two lambdas I showed:

```scheme
(arguments
 `(#:phases
   (modify-phases %standard-phases
     ;; ...
     (replace 'check
       ;; We bind "inputs" here, which is an association list of labels to
       ;; package store locations.
       (lambda* (#:key inputs #:allow-other-keys)
         ;; Set GUIX_LOCPATH to the "lib/locale" sub-directory of the
         ;; glibc-utf8-locales package in the inputs.
         (set-path-environment-variable
          "GUIX_LOCPATH" '("lib/locale")
          (list (assoc-ref inputs "glibc-utf8-locales")))
         (setenv "LANG" "en_US.UTF-8")

         ;; Add whatever directories contain executables that run_tut.sh
         ;; expects to be available on PATH.
         (setenv "PATH"
                 (string-append (getcwd)
                                "/dir/of/command/a:"
                                (getcwd)
                                "/dir/of/command/b:"
                                (getenv "PATH")))
         (with-directory-excursion "src/tutorial"
           (invoke "run_tut.sh"))))
     ;; ...
     )))
```

Hope that's a bit clearer.

mschilli87 commented:
Thanks for the explanation. Makes sense to me now. The last thing I don't know is how to obtain the `/dir/of/command`s: Those will be either provided by inputs or installed by the package itself, so basically I need the `bin` directory as created by the package in the store. If I run `guix environment mirdeep2` all I need to do is `cd "tutoral-dir" && "run_tut.sh"`. So I guess a shell would be required as well though it is not listed as input. Can I declare inputs for testing only?

mschilli87 commented:
Working on https://github.com/BIMSBbioinfo/guix-bimsb/pull/6 I remembered this and finally took the time to address most of the feedback.

The one thing still missing is the test pahse. For this I need to repeat my previous question:

Can I declare inputs for testing only?
The tutorial requires `bowtie-build` to be available but in principle a user might download an index and just run miRDeep2, so I'd prefer not to force installation `bowtie1` alongside `mirdeep2`. Also, the tutorial is implemented as a shell script. If we want to control this as well, a shell needs to be installed in the test environment as well.

To manually run the test, I use `guix environment --ad-hoc mirdeep2 bowtie1 bash-minimal` and then run

```sh
cd <mirdeep2_source_dir>/tutorial_dir && bash run_tut.sh
```

Any more pointers would be much appreciated so I can hopefully get this done (finally).

mschilli87 committed:
tainted: mirdeep2: Add test (tutorial)

Since the package definition had to practically re-implement the entire
installation procedure, the tutorial is run post-install (using the
patched and wrapped outputs).

The tutorial script 'handles' errors by printing an error message (incl.
the return value of the failed command) to stdout [sic]. Then, it
continues running as if nothing happens (potentially causing and
reporting further errors) and always returns zero in the end. Thus,
stdout is scanned for such error messages to determine failure or
success of the test.
Additional error may occur that are not handled by the tutorial script.
These result in error messages to stderr. Thus, stderr is combined with
stdout before scanning for error messages.
Scanning the tutorial script's output for error messages 'consumes' it,
such that it won't show in the Guix build log anymore. This, in case of
an error, understanding its nature would be harder than necessary.
Once option would be to report the actual error message found, but
additional context might be required. Instead of hard coding a fixed
number of pre-error lines to include in the build log (by replacing
`grep`'s `-q` flag by a `-B` one), the entire output of the tutorial
script (up to the error) is always included in the build log: This is
achieved by copying the (combined) output to stdout *and* stderr. This
way, the stdout copy can be consumed for the downstream checks but the
stderr one remains visible in the log.

mschilli87 commented:
@rekado: I think I finally did it. I'd consider this ready to merge. Thank you for all your help leading here and yet to come.

mschilli87 commented:
Rebased on latest master.

rekado commented on bimsb/packages/tainted.scm:
```diff
@@ -245,7 +244,36 @@ deviation (SD) plots, coefficient of variation (CV) plots.")
+               ;; Run script in bash to processs output for error detection:
+               (invoke
+                 "bash" "-c"
+                 (string-append
```
This is not a big problem, obviously, but it would be nicer not to use `string-append` and just use a single string.  You can escape line breaks with `\`.

rekado commented:
Thanks for working on it with so much dedication!

mschilli87 commented:
Thank you for the guidance! :slightly_smiling_face:
